R/meltAssay.R

```diff
@@ -0,0 +1,152 @@
+#' @title meltAssay
+#'
+#' @description Converts a \code{\link[=TreeSummarizedExperiment-class]{TreeSummarizedExperiment}} object into a
+#' long data.frame which can be used for \code{\link[tidyverse]{tidyverse}}-tools.
+#'
+#' @param x a numeric matrix or a \code{\link[=TreeSummarizedExperiment-class]{TreeSummarizedExperiment}}
+#'          object containing a tree.
+#' @param abund_values Must be one of the values returned by assayNames.
+#' @param add_col_data Choice of colData columns to add. Default="none".
+#' @param ... optional arguments not used.
+#' @return A data.frame tbl_df.
+#' @examples
+#' \dontrun{
+#' data(GlobalPatterns)
+#' data(GlobalPatterns, package = "MicrobiomeExperiment")
+#' se <- GlobalPatterns
+#' se.rel <- relAbundanceCounts(se)
+#' molten_se <- meltAssay(se.rel,
+#'   add_col_data = colnames(colData(se.rel)),
+#'   abund_values = "counts"
+#' )
+#' head(molten_se)
+#' }
+#' @importFrom SummarizedExperiment assayNames rowData rowData<-
+#' @importFrom tibble rownames_to_column
+#' @importFrom tidyr pivot_longer
+#' @importFrom dplyr left_join mutate_if %>%
+#' @keywords Utilities
+#' @export
+setGeneric("meltAssay",
+  signature = "x",
+  function(x, add_col_data = "none", abund_values = "counts", ...) {
```

Comment:
I think it would be good, if `add_col_data` would accept the following values:

- NULL so no data gets added.
- character values to select certain columns
- TRUE to select all columns.

It could be the case, that a column is named `none` or `all`, which then couldn't be added to the molten data. The same is also true for taxonomic data. I think we need a argument for that as well. Maybe it is also good to use a general term like `add_row_data`. What do you think?

Reply:
I made changes to `add_col_data` options. Regarding `add_row_data`, I am not sure if this option will be of use. This is mostly based on my experience with microbiome data. The `rowData` is one of the most useful information for interpretation and data visualization. `phyloseq` also returns `tax_table` by default. Additionally, the hierarchical nature of information with often *missing* classification at various levels hampers interpretation and users would always keep all information at all taxonomic levels.

Reply:
I don't see this as a major problem. Maybe this gets us back to a conceptual discussion. `meltAssay` is for me the main function, the workhorse so to say. However, I don't see that this as a strictly user facing functions.

Think about it this way: What task would a user have towards the object and its data? 
1. Give me the abundance data with all/some/these taxonomic information.
2. give me the abundance data with with certain covariate information.
`melting` is a really technical term, which isn't even used in the tidyverse anymore (they now use `pivot_longer`) and is certainly not a term associated by default with tasks outlined above.

Therefore, at least in my mind, we will have a function `getAbundanceData` or something, which will have different default and will get the task done.

So this is why I am strongly arguing, that we keep our options open.

`meltAssay` must be able to add data, but only, when required. If we want to reuse `meltAssay` for some plotting or other internal functions, I don't want to right the function again, just because `meltAssay` gives me data, I didn't ask for. That will blow the memory consumption at the most unfortunate moment, aka. the moment where you are running low on memory. And then this discussion we are having matters.